File: collector/core/runner.go
package core

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"
)

// boundSource is a configured source with its poll interval and resolved
// destinations.
type boundSource struct {
	name         string
	typeName     string
	interval     time.Duration
	source       Source
	destinations []boundDestination
}

type boundDestination struct {
	name        string
	destination Destination
}

// Runner owns all configured sources and destinations and schedules
// collection.
type Runner struct {
	log          *slog.Logger
	sources      []boundSource
	destinations []boundDestination
}

// NewRunner instantiates every configured destination and source. A source
// with no explicit `destinations` list sends to all destinations.
func NewRunner(conf *Config, log *slog.Logger) (*Runner, error) {
	r := &Runner{log: log}

	destsByName := map[string]boundDestination{}
	for _, dc := range conf.Destinations {
		dest, err := NewDestination(dc.Type, dc.Config)
		if err != nil {
			return nil, fmt.Errorf("destination %q: %w", dc.Name, err)
		}
		bound := boundDestination{name: dc.Name, destination: dest}
		destsByName[dc.Name] = bound
		r.destinations = append(r.destinations, bound)
	}

	for _, sc := range conf.Sources {
		src, err := NewSource(sc.Type, sc.Config)
		if err != nil {
			return nil, fmt.Errorf("source %q: %w", sc.Name, err)
		}

		intervalStr := sc.Interval
		if intervalStr == "" {
			intervalStr = conf.DefaultInterval
		}
		interval, err := time.ParseDuration(intervalStr)
		if err != nil {
			return nil, fmt.Errorf("source %q: invalid interval %q: %w", sc.Name, intervalStr, err)
		}

		bound := boundSource{
			name:     sc.Name,
			typeName: sc.Type,
			interval: interval,
			source:   src,
		}
		if len(sc.Destinations) == 0 {
			bound.destinations = r.destinations
		} else {
			for _, dn := range sc.Destinations {
				bound.destinations = append(bound.destinations, destsByName[dn])
			}
		}
		r.sources = append(r.sources, bound)
	}
	return r, nil
}

// Run collects from every source on its interval until ctx is cancelled.
// Sources with a 0 interval collect exactly once. Run returns once all
// source loops have stopped and destinations are closed.
func (r *Runner) Run(ctx context.Context) error {
	var wg sync.WaitGroup
	for i := range r.sources {
		src := &r.sources[i]
		wg.Add(1)
		go func() {
			defer wg.Done()
			r.runSource(ctx, src)
		}()
	}
	wg.Wait()

	closeCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	for _, d := range r.destinations {
		if err := d.destination.Close(closeCtx); err != nil {
			r.log.Error("failed to close destination", "destination", d.name, "error", err)
		}
	}
	return nil
}

func (r *Runner) runSource(ctx context.Context, src *boundSource) {
	log := r.log.With("source", src.name, "type", src.typeName)
	for {
		start := time.Now()
		batch, err := src.source.Collect(ctx)
		if err != nil {
			if ctx.Err() != nil {
				return
			}
			log.Error("collection failed", "error", err)
		} else {
			log.Info("collected", "messages", len(batch), "duration", time.Since(start).Round(time.Millisecond))
			for i := range batch {
				if batch[i].Meta == nil {
					batch[i].Meta = map[string]string{}
				}
				batch[i].Meta["source"] = src.name
				batch[i].Meta["source_type"] = src.typeName
			}
			for _, d := range src.destinations {
				if len(batch) == 0 {
					continue
				}
				if err := d.destination.Write(ctx, batch); err != nil {
					if ctx.Err() != nil {
						return
					}
					log.Error("write failed", "destination", d.name, "error", err)
				}
			}
		}

		if src.interval <= 0 {
			log.Info("interval is 0, source finished after one collection")
			return
		}
		select {
		case <-time.After(src.interval):
		case <-ctx.Done():
			return
		}
	}
}

File: collector-rs/src/runner.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use anyhow::{Context, Result};
use log::{error, info};
use tokio::sync::watch;

use crate::config::{parse_duration, Config};
use crate::registry::{Destination, Registry, Source};

struct BoundSource {
    name: String,
    type_name: String,
    interval: Duration,
    source: Box<dyn Source>,
    destinations: Vec<(String, Arc<dyn Destination>)>,
}

/// Owns all configured sources and destinations and schedules collection.
pub struct Runner {
    sources: Vec<BoundSource>,
    destinations: Vec<(String, Arc<dyn Destination>)>,
}

impl Runner {
    /// Instantiates every configured destination and source. A source with
    /// no explicit `destinations` list sends to all destinations.
    pub fn new(conf: &Config, registry: &Registry) -> Result<Self> {
        let mut destinations: Vec<(String, Arc<dyn Destination>)> = Vec::new();
        for d in &conf.destinations {
            let dest = registry
                .new_destination(&d.type_name, &d.config)
                .with_context(|| format!("destination {:?}", d.name))?;
            destinations.push((d.name.clone(), Arc::from(dest)));
        }

        let mut sources = Vec::new();
        for s in &conf.sources {
            let source = registry
                .new_source(&s.type_name, &s.config)
                .with_context(|| format!("source {:?}", s.name))?;
            let interval_str = s.interval.as_deref().unwrap_or(&conf.default_interval);
            let interval = parse_duration(interval_str)?;

            let bound_dests = if s.destinations.is_empty() {
                destinations.clone()
            } else {
                destinations
                    .iter()
                    .filter(|(name, _)| s.destinations.contains(name))
                    .cloned()
                    .collect()
            };

            sources.push(BoundSource {
                name: s.name.clone(),
                type_name: s.type_name.clone(),
                interval,
                source,
                destinations: bound_dests,
            });
        }

        Ok(Self {
            sources,
            destinations,
        })
    }

    /// Collects from every source on its interval until `shutdown` fires.
    /// Sources with a 0 interval collect exactly once.
    pub async fn run(self, shutdown: watch::Receiver<bool>) {
        let mut handles = Vec::new();
        for src in self.sources {
            let mut shutdown = shutdown.clone();
            handles.push(tokio::spawn(async move {
                run_source(src, &mut shutdown).await;
            }));
        }
        for handle in handles {
            let _ = handle.await;
        }
        for (name, dest) in &self.destinations {
            if let Err(err) = dest.close().await {
                error!("failed to close destination {name:?}: {err:#}");
            }
        }
    }
}

async fn run_source(src: BoundSource, shutdown: &mut watch::Receiver<bool>) {
    loop {
        let start = Instant::now();
        match src.source.collect().await {
            Ok(mut batch) => {
                info!(
                    "source {:?} collected {} messages in {:?}",
                    src.name,
                    batch.len(),
                    start.elapsed()
                );
                for msg in &mut batch {
                    msg.meta.insert("source".to_string(), src.name.clone());
                    msg.meta
                        .insert("source_type".to_string(), src.type_name.clone());
                }
                if !batch.is_empty() {
                    for (dest_name, dest) in &src.destinations {
                        if let Err(err) = dest.write(&batch).await {
                            error!(
                                "source {:?} write to destination {dest_name:?} failed: {err:#}",
                                src.name
                            );
                        }
                    }
                }
            }
            Err(err) => error!("source {:?} collection failed: {err:#}", src.name),
        }

        if src.interval.is_zero() {
            info!("source {:?} finished after one collection", src.name);
            return;
        }
        tokio::select! {
            _ = tokio::time::sleep(src.interval) => {}
            _ = shutdown.changed() => return,
        }
    }
}
